File: leetcode/142. Linked List Cycle II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        slow = head
        fast = head
        
        while fast != None and fast.next != None:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                break
        
        if fast == None or fast.next == None:
            return None

        slow2 = head
        while slow2 != slow:
            slow2 = slow2.next
            slow = slow.next

        return slow

File: leetcode/142. Linked List Cycle II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        ListNode* fast = head;
        ListNode* slow = head;
        
        while(fast != nullptr && fast->next != nullptr){
            slow = slow->next;
            fast = fast->next->next;
            if(slow == fast){
                break;
            }
        }
        
        if(fast == nullptr || fast->next == nullptr){
            return nullptr;
        }
        
        ListNode* slow2 = head;
        while(slow2 != slow){
            slow = slow->next;
            slow2 = slow2->next;
        }
        return slow;
    }
};
